## 분류 모델 수행하기

### 2.1 데이터 탐색하기

#### 데이터 가져오기

In [33]:
import pandas as pd

In [74]:
X_train = pd.read_csv('./bigData-main/titanic_x_train.csv', encoding='cp949')
X_test  = pd.read_csv('./bigData-main/titanic_x_test.csv', encoding='cp949')
y_train = pd.read_csv('./bigData-main/titanic_y_train.csv', encoding='cp949')

print(X_train.head()) # NaN 있음
print(X_test.head())  # NaN 있음
print(y_train.head()) # PassengerId, Survived

   PassengerId  티켓등급                                               승객이름  \
0            1     3                            Braund, Mr. Owen Harris   
1            2     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3     3                             Heikkinen, Miss. Laina   
3            4     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            5     3                           Allen, Mr. William Henry   

       성별    나이  형제자매배우자수  부모자식수              티켓번호     운임요금  객실번호 선착장  
0    male  22.0         1      0         A/5 21171   7.2500   NaN   S  
1  female  38.0         1      0          PC 17599  71.2833   C85   C  
2  female  26.0         0      0  STON/O2. 3101282   7.9250   NaN   S  
3  female  35.0         1      0            113803  53.1000  C123   S  
4    male  35.0         0      0            373450   8.0500   NaN   S  
   PassengerId  티켓등급                                          승객이름      성별  \
0          892     3                   

#### 행/열 확인하기

In [75]:
print(X_train.shape, X_test.shape, y_train.shape)

(891, 11) (418, 11) (891, 2)


#### 요약정보 확인하기

- 나이, 객실번호, 선착장 컬럼 결측치 확인
- object : 승객이름, 성별, 티켓번호, 객실번호, 선착장 > 범주형 변수인지 확인 필요

In [76]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   티켓등급         891 non-null    int64  
 2   승객이름         891 non-null    object 
 3   성별           891 non-null    object 
 4   나이           714 non-null    float64
 5   형제자매배우자수     891 non-null    int64  
 6   부모자식수        891 non-null    int64  
 7   티켓번호         891 non-null    object 
 8   운임요금         891 non-null    float64
 9   객실번호         204 non-null    object 
 10  선착장          889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


In [77]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   티켓등급         418 non-null    int64  
 2   승객이름         418 non-null    object 
 3   성별           418 non-null    object 
 4   나이           332 non-null    float64
 5   형제자매배우자수     418 non-null    int64  
 6   부모자식수        418 non-null    int64  
 7   티켓번호         418 non-null    object 
 8   운임요금         417 non-null    float64
 9   객실번호         91 non-null     object 
 10  선착장          418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [78]:
y_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   PassengerId  891 non-null    int64
 1   Survived     891 non-null    int64
dtypes: int64(2)
memory usage: 14.0 KB


In [79]:
X_train.select_dtypes(include = object).columns

Index(['승객이름', '성별', '티켓번호', '객실번호', '선착장'], dtype='object')

In [80]:
print('성별     - ', X_train['성별'].unique().size)
print(X_train['성별'].unique(),'\n')

print('선착장   - ', X_train['선착장'].unique().size)
print(X_train['선착장'].unique(), '\n')

print('승객번호 - ', X_train['승객이름'].nunique())
print('티켓번호 - ', X_train['티켓번호'].nunique())
print('객실번호 - ', X_train['객실번호'].nunique())

성별     -  2
['male' 'female'] 

선착장   -  4
['S' 'C' 'Q' nan] 

승객번호 -  891
티켓번호 -  681
객실번호 -  147


- 성별, 선착장 : 10개 이하의 값 종류 > 범주형 데이터
- 승객이름, 티켓번호, 객실번호 : 수백 개의 값을 가짐 > 범주를 갖지 않는 문자열 값

#### 기초 통계량 확인하기

In [81]:
print(X_train.describe())

       PassengerId        티켓등급          나이    형제자매배우자수       부모자식수        운임요금
count   891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean    446.000000    2.308642   29.699118    0.523008    0.381594   32.204208
std     257.353842    0.836071   14.526497    1.102743    0.806057   49.693429
min       1.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    2.000000   20.125000    0.000000    0.000000    7.910400
50%     446.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    3.000000   38.000000    1.000000    0.000000   31.000000
max     891.000000    3.000000   80.000000    8.000000    6.000000  512.329200


#### 독립변수 - 종속변수 관계 확인하기

- 독립변수의 영향도와 추세 파악

In [82]:
# X_train, y_train 을 세로 방향으로 결합 후, data 변수에 저장
data = pd.concat([X_train, y_train], axis=1)

In [83]:
# '성별' 컬럼에 따라 Survived 평균값 계산
print(data.groupby('성별')['Survived'].mean())
print('\n>>>> 여성이 남성보다 생존율이 높음 확인')

성별
female    0.742038
male      0.188908
Name: Survived, dtype: float64

>>>> 여성이 남성보다 생존율이 높음 확인


In [84]:
# '티켓등급' 컬럼에 따라 Survived 평균값 계산
print(data.groupby('티켓등급')['Survived'].mean())
print('\n>>>> 티켓등급이 높을수록 생존율이 높음 예상')

티켓등급
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

>>>> 티켓등급이 높을수록 생존율이 높음 예상


In [85]:
# '선착장' 컬럼에 따라 Survived 평균값 계산
print(data.groupby('선착장')['Survived'].mean())
print('\n>>>> C 선착장에 탑승한 승객의 생존율이 가장 높음 확인')

선착장
C    0.553571
Q    0.389610
S    0.336957
Name: Survived, dtype: float64

>>>> C 선착장에 탑승한 승객의 생존율이 가장 높음 확인


'성별', '티켓등급', '선착장' 컬럼들은 최종적으로 분류해야 하는 Survived 컬럼과 명확한 관계가 있으므로, 데이터 분석 대상에 반드시 포함!

### 데이터 전처리

#### 불필요한 컬럼 삭제

In [86]:
# 테스트 데이터의 PassengerId 값은 x_test_passenger_id 변수에 저장
x_test_passenger_id = X_test.PassengerId

In [87]:
# X_train, y_train, X_test 에서 PassengerId 컬럼 삭제
X_train = X_train.drop(columns = ['PassengerId'])
X_test  = X_test.drop(columns = ['PassengerId'])
y_train = y_train.drop(columns = ['PassengerId'])

In [88]:
print(y_train.head()) # PassengerId 컬럼 삭제 확인

   Survived
0         0
1         1
2         1
3         1
4         0


문자형 데이터인 '티켓번호' 칼럼은 681건의 중복 제거된 값을 가지고 있으므로, 이는 전체 대비 50%가 넘는 데이터로 분석 대상에서 삭제 처리

In [89]:
# print(X_train['티켓번호'].nunique())  > 681

# 학습 데이터 및 테스트 데이터에서 '티켓번호' 컬럼 삭제
X_train = X_train.drop(columns = ['티켓번호'])
X_test  = X_test.drop(columns = ['티켓번호'])

문자형 데이터인 '승객이름' 컬럼은 전체 데이터 건수만큼 서로 다른 값을 가지게 되기 때문에, 단순한 데이터 분석을 위해 삭제 처리 진행
- 이름을 구성하는 세가지 요소(성, 호칭, 이름)를 분리하여 패턴화할 수도 있으나, 간단한 데이터 분석을 위해 편의상 삭제

In [90]:
# 학습 데이터와 테스트 데이터에서 '승객이름' 컬럼 삭제
X_train = X_train.drop(columns = ['승객이름'])
X_test  = X_test.drop(columns = ['승객이름'])

### 결측치 처리

In [91]:
# '나이', '객실번호', '선착장' 컬럼의 결측치 개수 세기
print(X_train['나이'].isnull().sum())
print(X_train['객실번호'].isnull().sum())
print(X_train['선착장'].isnull().sum())

177
687
2


나이 컬럼과 Survived 컬럼의 상관성 확인
- 상관관계가 약 -0.08 로 매우 낮아 분석대상에서 제외

In [92]:
print(data[['나이', 'Survived']].corr())

                나이  Survived
나이        1.000000 -0.077221
Survived -0.077221  1.000000


In [93]:
# '나이' 컬럼 삭제
X_train = X_train.drop(columns = ['나이'])
X_test  = X_test.drop(columns = ['나이'])

객실번호 결측치 687건, nunique = 148건 으로 대부분의 데이터가 서로 상이한 값이고 결측치 처리 어려우므로 분석대상에서 제외

In [94]:
# '객실번호' 컬럼 삭제
X_train = X_train.drop(columns = ['객실번호'])
X_test  = X_test.drop(columns = ['객실번호'])

선착장 컬럼은 결측치 2건으로, 가장 많은 비율을 차지하는 값으로 적용 > S 값으로 대치

In [95]:
print(X_train.groupby('선착장')['선착장'].count())

선착장
C    168
Q     77
S    644
Name: 선착장, dtype: int64


In [96]:
X_train['선착장'] = X_train['선착장'].fillna('S')

print(X_train['선착장'].isnull().sum())

0


In [97]:
print(X_test['선착장'].isnull().sum())  # 테스트 데이터 결측치 0건으로 pass

0


#### 범주형 변수 인코딩

- '성별' 컬럼 연속형 변수로 변환 >> male : 0, female : 1

In [98]:
X_train['성별'] = X_train['성별'].replace('male', 0).replace('female', 1)
X_test['성별']  = X_test['성별'].replace('male', 0).replace('female', 1)

- '선착장' 컬럼 연속형 변수로 변환 >> 세가지 종류의 데이터로 구성, 원핫인코딩 수행 >> pd.get_dummies() 사용

In [99]:
선착장_dummy = pd.get_dummies(X_train['선착장'], drop_first = True).rename(columns = {'Q' : '선착장Q', 'S' : '선착장S'})

# 기존 X_train 우측에 변수 덧붙여 X_train 에 다시 저장
X_train = pd.concat([X_train, 선착장_dummy], axis=1)

# X_train 의 상위 5개 행 확인
print(X_train.head())

   티켓등급  성별  형제자매배우자수  부모자식수     운임요금 선착장  선착장Q  선착장S
0     3   0         1      0   7.2500   S     0     1
1     1   1         1      0  71.2833   C     0     0
2     3   1         0      0   7.9250   S     0     1
3     1   1         1      0  53.1000   S     0     1
4     3   0         0      0   8.0500   S     0     1


In [100]:
# 학습데이터에서 '선착장' 컬럼 삭제
X_train = X_train.drop(columns = ['선착장'])

테스트 데이터도 동일하게 진행 : '선착장' 컬럼 원핫인코딩 수행 > 기존 컬럼 삭제

In [101]:
선착장_dummy = pd.get_dummies(X_test['선착장'], drop_first = True).rename(columns = {'Q':'선착장Q', 'S':'선착장S'})
X_test = pd.concat([X_test, 선착장_dummy], axis=1)
X_test = X_test.drop(columns = ['선착장'])
X_test.head()

티켓등급  성별  형제자매배우자수  부모자식수     운임요금  선착장Q  선착장S
0     3   0         0      0   7.8292     1     0
1     3   1         1      0   7.0000     0     1
2     2   0         0      0   9.6875     1     0
3     3   0         0      0   8.6625     0     1
4     3   1         1      1  12.2875     0     1

#### 파생변수 만들기

- '형재자매배우자수' 컬럼과 '부모자식수' 컬럼으로 '가족수' 라는 파생변수 생성

In [105]:
X_train['가족수'] = X_train['형제자매배우자수'] + X_train['부모자식수']
print(X_train[['형제자매배우자수', '부모자식수', '가족수']].head(10))

   형제자매배우자수  부모자식수  가족수
0         1      0    1
1         1      0    1
2         0      0    0
3         1      0    1
4         0      0    0
5         0      0    0
6         0      0    0
7         3      1    4
8         0      2    2
9         1      0    1


기존 2개의 컬럼 (형제자매배우자수, 부모자식수) 삭제

In [106]:
X_train = X_train.drop(columns = ['형제자매배우자수', '부모자식수'])

테스트 데이터도 동일하게 진행

In [108]:
X_test['가족수'] = X_test['형제자매배우자수'] + X_test['부모자식수']
X_test = X_test.drop(columns = ['형제자매배우자수', '부모자식수'])
X_test.head()

티켓등급  성별     운임요금  선착장Q  선착장S  가족수
0     3   0   7.8292     1     0    0
1     3   1   7.0000     0     1    1
2     2   0   9.6875     1     0    0
3     3   0   8.6625     0     1    0
4     3   1  12.2875     0     1    2

### 2.3 학습 및 평가

#### 데이터 분리

In [109]:
from sklearn.model_selection import train_test_split

X_TRAIN, X_VAL, Y_TRAIN, Y_VAL = train_test_split(X_train, y_train, test_size = .2, random_state = 42)

print(X_TRAIN.shape, X_VAL.shape, Y_TRAIN.shape, Y_VAL.shape)

(712, 6) (179, 6) (712, 1) (179, 1)


#### 데이터 학습 및 하이퍼 파라미터 튜닝

In [111]:
from xgboost import XGBClassifier

#### [1] XGBClassifier 기본값으로 모델 학습

In [112]:
model1 = XGBClassifier(eval_metric = 'error', random_state = 42) # eval_metric : 분류 모델로 학습을 수행하는 경우 작성하는 대표값
model1.fit(X_TRAIN, Y_TRAIN)

C:\Users\dltmf\anaconda3\envs\tensorEvn\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
C:\Users\dltmf\anaconda3\envs\tensorEvn\lib\site-packages\sklearn\preprocessing\_label.py:98: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dltmf\anaconda3\envs\tensorEvn\lib\site-packages\sklearn\preprocessing\_label.py:133: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel(

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
              eval_metric='error', gamma=0, gpu_id=-1, importance_type=None,
              interaction_constraints='', learning_rate=0.300000012,
              max_delta_step=0, max_depth=6, min_child_weight=1, missing=nan,
              monotone_constraints='()', n_estimators=100, n_jobs=8,
              num_parallel_tree=1, predictor='auto', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

#### [2] 하이퍼 파라미터 주입하여 모델 학습

- 대표적인 하이퍼 파라미터 : n_estimators, max_depth

In [113]:
model2 = XGBClassifier(n_estimators = 100, max_depth = 5, eval_metric = 'error', random_state = 42)
model2.fit(X_TRAIN, Y_TRAIN)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
              eval_metric='error', gamma=0, gpu_id=-1, importance_type=None,
              interaction_constraints='', learning_rate=0.300000012,
              max_delta_step=0, max_depth=5, min_child_weight=1, missing=nan,
              monotone_constraints='()', n_estimators=100, n_jobs=8,
              num_parallel_tree=1, predictor='auto', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

#### 결과 예측하기

In [120]:
y_pred1 = pd.DataFrame(model1.predict(X_test), columns = ['Survived'])
y_pred2 = pd.DataFrame(model2.predict(X_test), columns = ['Survived'])

print(pd.DataFrame(y_pred1).head(10))

   Survived
0         0
1         1
2         0
3         0
4         1
5         0
6         0
7         0
8         1
9         0


- 만약 승객이 사망할 확률을 구해야 하는 경우 : pd.DataFrame(model.predict_proba(x_test))[0]
- 승객이 생존할 확률을 구해야 하는 경우 : pd.DataFrame(model_predict_proba(x_test))[1]

ROC 평가지표 계산하기 위해 X_VAL 검증용 세트에 대응되는 Y_VAL 값 예측

In [117]:
Y_VAL_PRED1 = pd.DataFrame(model1.predict(X_VAL), columns = ['Survived'])
Y_VAL_PRED2 = pd.DataFrame(model2.predict(X_VAL), columns = ['Survived'])

#### 모델 평가하기

- 채점 기준인 roc_auc_score 함수 이용하여 모델 평가 수행

In [121]:
from sklearn.metrics import roc_auc_score

In [123]:
# 평가지표 결과 비교
print(roc_auc_score(Y_VAL, Y_VAL_PRED1))
print(roc_auc_score(Y_VAL, Y_VAL_PRED2))  # 두번째 모델을 최종 모델로 선정

0.7866795366795367
0.7942084942084942


### 2.4 결과 제출하기

- 최종데이터 : PassengerId 변수와 y_test 세트의 예측값(y_pred2) 함께 출력

In [129]:
final = pd.concat([x_test_passenger_id, y_pred2], axis=1)
final.to_csv('data/12345.csv', index=False) # 행번호 저장하지 않으므로 index=False 필수

In [131]:
# 저장된 파일 재확인
pd.read_csv('data/12345.csv').head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1